### ISSUE TITLE
Different prediction on same state in Stablebaseline3

### ISSUE THREAD

**Issue by sharvit-kesem** [AI4Finance-Foundation/FinRL#362]
Hey all, I used the Compare_ElegantRL_RLlib_Stablebaseline3 as inspiration, but noticed something that I don't understand, nice if someone can explain to me if I'm doing something wrong or the logic.
So basically, I trained Stablebaseline3/PPO/env_stocktrading_np and then backtest it, I get different actions from the model predictions for different runs(I use if model exists do not retrain). The if_train is false and I use the DRL_prediction_load_from_file method.
But I would get different first actions in different runs. If I use the Elegant it works as expected though.
Can someone help me understand, what am I missing here?

**Comment by trungtv** [comment]
You should set deterministic = True for prediction in the models.py.
`action, _states = model.predict(test_obs, deterministic=True)`

**Comment by hokhay** [comment]
> 

Why this is not a default setting for prediction?

**Comment by Athe-kunal** [comment]
In my use case, where I am trying to tune the hyperparameters for the validation data, deterministic = False makes sense because I want to test robust tuning. But yeah we can pass the deterministic bool value in DRL_prediction.

**Comment by hokhay** [comment]
> In my use case, where I am trying to tune the hyperparameters for the validation data, deterministic = False makes sense because I want to test robust tuning. But yeah we can pass the deterministic bool value in DRL_prediction.

Thanks for your reply, Athe. Sorry for my ignorance but I am still confused. When `deterministic = False`, the agent actions are taken as stochastic and this essentially introduces noise to the actions. I understand this noise is good for training phase as it makes the model more robust. 

However I do not understand why this is good for validation/prediction phase where the predicted result will not be feed backed to the model and so noise is not doing any good to the model. Controversially, a consist result is needed to determine the performance of the trained model. Please correct me if I am wrong.

Could you share more about your experience please?

Jason

**Comment by Athe-kunal** [comment]
I don't have strong reasoning. Actually, when I do hyperparameter tuning, I split my dataset into train -- val -- test. Train on train set and validate on Val set. So for train and validation, I let deterministic = False and for the test set, I do deterministic = True. So in the process, I tune the hyperparameters when the agent is behaving as stochastically as possible. Maybe I am getting it wrong. BUT for the test set, you should do deterministic=True.
